### ISSUE TITLE
Usage with React/JSX [question/feature request]

### ISSUE THREAD

**Issue by suprMax** [jashkenas/coffeescript#4331]
We're using React and https://github.com/jsdf/coffee-react in our project to allow CJSX syntax to function. It seems unfortunately that author gave up on it. I see no alternatives to compile the "xml tags" part otherwise. What is the recommended way to do this? Facebook pushes hard with Babel, while there seems to be a lack of any documentation in CS community regarding React.

I understand this is not a part of JavaScript and therefore is not being supported by CS. That being said, Babel supports a list of presets that allows for easy transformation of jsx to javascript. So does TypeScript: http://www.typescriptlang.org/docs/handbook/jsx.html

It would've been great to have one of:
 - Support for presets to have your coffee "reactified", and perhaps other features added on top. Seems there is a lot of people requesting "feature X" while others don't need it.
 - Support to somehow magically pair it with babel transform perhaps, to have coffee->babel->valid js?
 - Target this use case in the Readme

Perhaps this is one of "just use backtick" issues. If so, please share your success stories 👍 

TLDR: want to use React with CoffeeScript. Is there any support for this?

**Comment by vendethiel** [comment]
> Perhaps this is one of "just use backtick" issues. If so, please share your success stories 👍

Sure. Or, since it's all functions anyway, just call the functions?

**Comment by suprMax** [comment]
Technically correct but writing code like that sucks. Kind of defeats the purpose of having coffee script and then that

**Comment by rafgugi** [comment]
whats wrong with this? im still using standard coffescript to write react

```coffeescript
dom = React.createElement

Button = React.createClass
  displayName: 'Button'

  propTypes: ->
    id: React.PropTypes.string
    className: React.PropTypes.string
    onClick: React.PropTypes.func

  getDefaultProps: ->
    id: ''
    className: 'btn-default'
    onClick: -> console.log "Button Clicked"

  render: ->
    dom 'button', 
      type: 'button'
      id: @props.id
      className: 'btn ' + @props.className
      onClick: @props.onClick
      @props.children

module.exports = Button
```

**Comment by suprMax** [comment]
Nothing wrong. It's just not as terse as JSX syntax.

**Comment by GeoffreyBooth** [comment]
I’m planning on building a Meteor + React + CoffeeScript app soon, and I was thinking of following this example: https://github.com/TheAncientGoat/mantra-sample-blog-coffee. Discussion [here](https://forums.meteor.com/t/coffeescript-and-meteor-1-3-how-to-use-modules-example-todos-app/20622/4?u=geoffreybooth). I’m not sure how the Jade-like syntax works, but I’m hoping to figure it out 😄

**Comment by williamstein** [comment]
> TLDR: want to use React with CoffeeScript. Is there any support for this?

Despite https://github.com/jsdf/coffee-react being deprecated, we've been using it extensively on a [massive coffeescript/react codebase](https://github.com/sagemathinc/smc/tree/master/src/smc-webapp) with no problems for a while.  
We have no plans to stop using it, and if any issues arise we (at SageMath, Inc.) will likely fork coffee-react, dig into the code (it's a total of less than 500 lines of coffeescript), and fix issues.  

So https://github.com/jsdf/coffee-react may be deprecated by its original author, but that definitely doesn't mean it's dead.

**Comment by suprMax** [comment]
@williamstein it's been deprecated for bit over a week, that's why no problems yet. I would assume next React release might have breaking changes (again) that would break compiler (again). After that it's uncharted waters. Would be so nice if CJSX syntax gets adopted as part of CoffeeScript.

**Comment by williamstein** [comment]
Update to my comment: Of course https://github.com/jsdf/coffee-react-transform is also where the real action s, and it's another 1000+ lines of code.  But it seems like easily readable code.

Looking at https://github.com/jsdf/coffee-react-transform/commits/master, it looks like all the hard recent work was PR's from people other than the author who decided to deprecate?  That's a very good sign regarding the project having a future.

**Comment by williamstein** [comment]
I've also forked and "de-deprecated": coffee-react-transform:   https://github.com/sagemathinc/coffee-react-transform

Not sure what to do about the npm module names.  We did ask the author of  coffee-react-transform to transfer the github repos, but he refused.

**Comment by suprMax** [comment]
Not very nice of him. Doesn't want to maintain, yet doesn't want to transfer either. I suppose renaming it to "cjsx-*" would be a good move?

**Comment by williamstein** [comment]
>  I suppose renaming it to "cjsx-*" would be a good move?

That does make a lot of sense, especialy for the npm modules.   

Do you want me to add you to our sagemathinc/coffee-react* repos?  I'm happy to do so.

**Comment by jashkenas** [comment]
CoffeeScript isn't going to add special-case support for JSX syntax.

I do think it would be *very* interesting to work on a language that unifies the web platform, but this ain't it. Maybe check out Imba?

**Comment by williamstein** [comment]
By Imba, I think @jashkenas means https://github.com/somebee/imba....  @jashkenas thanks for the reference (and for the beautiful clean stable as a rock coffeescript, which I love!).

**Comment by suprMax** [comment]
Imba looks really cool!

@williamstein thanks for the offer. I don't think I have the capacity to actively maintain it at the moment, but I will ping you if that changes. We use CJSX at work and it's quite a significant amount of code, so we are invested in keeping it alive as long as we can.
